cEP-0029: Support TOML as a configuration format (#180)
coala/cEPs: cEP-0029.md

# Support TOML as a configuration format

| Metadata |                                           |
| -------- | ----------------------------------------- |
| cEP      | 29                                       |
| Version  | 1.0                                       |
| Title    | Support TOML as a configuration format     |
| Authors  | Prajwal M <mailto:prajwalmmath@gmail.com> |
| Status   | Implementation Due                        |
| Type     | Feature                                   |

## Abstract

This document describes how support for TOML will be implemented and integrated
with the existing configuration system. It also describes how other features that

> that complement the base idea will be implemented as part of this

Reviewer: A redundant 'that'

[GSoC project](https://summerofcode.withgoogle.com/projects/#6388671438127104).

## Introduction

Having a configuration file is essential to make full use of coala.
Having a standardised configuration format like TOML helps.
coala currently has an INI style configuration. INI style configurations
has certain limitations when it comes to custom sub-level parsing.
The LineParser and ConfParser becomes complex and error-prone.
A standardised configuration format allows developers to quickly
implement new features and will allow users to write configuration files

> in a format that they are already familiar with.

Reviewer: ```suggestion
in a format that they are probably more familiar with.
```
This sounds better to me

[all.a]
b = 10
c = 20
d = 30
```

Since aspects are settings within a section and since aspects
have their own group of settings, this syntax perfectly suits
aspects.

Aspects in coafile
```text
[all]
files = **
aspects = aspectname1, AspectName2 # case-insensitive
# defining an aspect's taste
aspectname1:aspect_taste = 80
# we can define subaspect taste through its parent
aspectname1:subaspect_taste = word1, word2, word3
```

is equivalent to

```toml
[all]
files = '**'
aspects = ['aspectname1', 'AspectName2'] # case-insensitive

[all.aspectname1]
aspect_taste = 80
subaspect_taste = ['word1', 'word2', 'word3']
```

or

```toml
[all]
files = '**'
aspects = ['aspectname1', 'AspectName2'] # case-insensitive
aspectname1.aspect_taste = 80
AspectName2.subaspect_taste = ['word1', 'word2', 'word3']
```
in TOML.

### Inheritance

Inheritance in coafile
```text
[all]
enabled = True
overridable = 2
ignore = vendor1/

[all.section1]
overridable = 3
ignore += vendor2/
other = some_value
```

is equivalent to

```toml
[all]
enabled = true
overridable = 2
ignore = 'vendor1/'

[section1]
inherits = 'all'
appends = 'ignore'
overrideable = 3
ignore = 'value2/'
other = 'some_value'
```

in TOML

> ### Trying to have syntax like coafile

Reviewer 2: No need, we will follow the TOML standard format in most cases because current users of TOML will get confused, we will customize only when it is necessary.

(Leave this section here, don't remove)

into `toml.load`. For this a new configuration flag
called `coala -T` shall be created. This will enable coala to
run on toml config files. Also creating such a flag is highly
beneficial. This way the flags -c , -F, -I and -s can also be
used with respect to toml.

To achieve this, I will refactor the `load_configuration` function.
I will not change the nature or number of function parameters
and return values. This way the behaviour of the function shall
remain the same. If `coala -T` is will call a function that
will provide toml sections and targets else it will provide normal
sections and targets. A function called `load_toml_config_file`
equivalent to `load_config_file` will be created.

### Implementation of TomlConfParser

The goal of this part is to get the OrderedDict from `toml.load`
and convert it into sections. For each table, the individual
key-value pairs have to converted to Settings. Finally these
settings have to grouped into a section. The concepts of
aspects and inheritance has to be accounted for. It's
function is similar to the present `ConfParser`.

This enables coala to run on configurations written in TOML
file.

> It should convet

Reviewer: ```suggestion
It should convert
```

                            ['*.py',
                             'coalib/**/*.py',
                             './coala',
                             'tests/**/*.py',
                             'docs/conf.py']),
                           ('ignore',
                            ['tests/bearlib/languages/documentation/documentation_extraction_testdata/*.py',
                             'tests/collecting/collectors_test_dir/bears/incorrect_bear.py',
                             'coalib/misc/Asyncio.py']),
                           ('max_line_length', 80),
                           ('use_spaces', True)]))])
```
into

```python
OrderedDict([('default',
              <Section object(aspects=None,
              contents=OrderedDict([('files', <Setting object(key='files', value='*.py, coalib/**/*.py, ./coala, tests/**/*.py, docs/conf.py', origin='/Users/prajwalm/Documents/coala/coala/.coafile', from_cli=False, to_append=False) at 0x1092cd780>),
              ('ignore', <Setting object(key='ignore', value='tests/bearlib/languages/documentation/documentation_extraction_testdata/*.py,\ntests/collecting/collectors_test_dir/bears/incorrect_bear.py,\ncoalib/misc/Asyncio.py,',

>               origin='/Users/prajwalm/Documents/coala/coala/.coafile', from_cli=False, to_append=False) at 0x1092cdbe0>),

Reviewer: I suggest you avoid using your own username 'prajwalm' here

              ('comment0', <Setting object(key='comment0', value='', origin='/Users/prajwalm/Documents/coala/coala/.coafile', from_cli=False, to_append=False) at 0x1092cdb70>),
              ('max_line_length', <Setting object(key='max_line_length', value='80', origin='/Users/prajwalm/Documents/coala/coala/.coafile', from_cli=False, to_append=False) at 0x1092cda20>),
              ('use_spaces', <Setting object(key='use_spaces', value='True', origin='/Users/prajwalm/Documents/coala/coala/.coafile', from_cli=False, to_append=False) at 0x1092cd898>)]), defaults=None,
              language=None, name='Default') at 0x108bdf550>)])
```

### Implementation of TomlConfWriter

The TomlConfWriter receives sections and it has to convert it into a OrderedDict
for `toml.dump`. This writes the sections to a TOML config file. It has a function
that is opposite to the TomlConfParser. It's
function is similar to the present `ConfWriter`.

This will enable users to use --save in TOML. It will also allow coala-quickstart
to generate TOML config files.

> It should convet

Reviewer: ```suggestion
It should convert
```

TOML.

New functions `write_toml_config` and `write_toml_info` will
be created which will perform the functions similar to existing
`write_coafile` and `write_info` functions but with respect to TOML.

Regarding green mode, I will create `write_toml_sections` function
similar to `write_sections` for TOML. Also the function
`generate_green_mode_sections` will be modified to accommodate
for TOML config file.

### Develop TOMLLintBear

coala-bears currently has bear for TOML called `TOMLBear.py` which
uses a external linter. But it requires Go to be installed.
So this bear can mainly be used only by Go users.

The task is to create `TOMLLintBear` in python. This way it can
be used by all users without any additional language requirements
as python comes installed in most computers.

The bear shall catch syntax errors in a TOML file. This shall be
done in 2 passes.
- In the first pass, the TOML file shall be validated per character
- In the second pass, the TOML file shall be validated per line

This approach is similar to the toml library.

### A repository of good config files to allow users to easily bootstrap their projects.

> 

Reviewer: Is this section unfinished or deliberately left empty?

Reviewer 3: If this project is carried, should the foundation of it -- i.e. adding some example config files -- be started within this project? I feel that some examples and community-provided configurations sould be present from the start.

# Support TOML as a configuration format

| Metadata |                                           |
| -------- | ----------------------------------------- |
| cEP      | 29                                       |
| Version  | 1.0                                       |
| Title    | Support TOML as a configuration format     |
| Authors  | Prajwal M <mailto:prajwalmmath@gmail.com> |
| Status   | Implementation Due                        |
| Type     | Feature                                   |

## Abstract

This document describes how support for TOML will be implemented and integrated
with the existing configuration system. It also describes how other features
that complement the base idea will be implemented as part of this
[GSoC project](https://summerofcode.withgoogle.com/projects/#6388671438127104).

## Introduction

Having a configuration file is essential to make full use of coala.

> Having a standardised configuration format like TOML helps.

Reviewer 4: standardised --> standardized . The same change on line 26 below

Author: US vs UK format

`write_coafile` and `write_info` functions but with respect to TOML.

Regarding green mode, I will create `write_toml_sections` function
similar to `write_sections` for TOML. Also the function
`generate_green_mode_sections` will be modified to accommodate
for TOML config file.

### Develop TOMLLintBear

coala-bears currently has bear for TOML called `TOMLBear.py` which
uses a external linter. But it requires Go to be installed.
So this bear can mainly be used only by Go users.

The task is to create `TOMLLintBear` in python. This way it can
be used by all users without any additional language requirements
as python comes installed in most computers.

The bear shall catch syntax errors in a TOML file. This shall be
done in 2 passes.
- In the first pass, the TOML file shall be validated per character
- In the second pass, the TOML file shall be validated per line

This approach is similar to the toml library. As part of this project
the bear shall be a validator and only check the validity of the TOML
file. In the future when style guidelines are created formatting
capability can be added.

### A repository of good config files to allow users to easily bootstrap their projects.

> 

Reviewer 5: I feel that this project is related but some example config files and some community-provided examples should be there from the start. Also, don't you think that this should be included within the TOML support itself, not as a separate project?

Author: @margobra8 I will provide example config files and documentation in TOML from the start to help users write configuration files in TOML like http://docs.coala.io/en/latest/Users/coafile.html.

`A repository of good config files to allow users to easily bootstrap their projects.` is about creating a separate repo similar to https://github.com/github/gitignore and https://github.com/moremoban/pypi-mobans to allow easy bootstrapping of projects.

Reviewer 5: @PrajwalM2212 the thing I'm trying to get to is if you actually consider generating another repo with templates as part of the initial requirements of the project or as future development or to be implemented later.

Author: As part of GSoC, another repo will be created with few templates as part of the initial requirements of the project.  A minimal version of `A repository of good config files` will be created during GSoC.
